KC7-Foundation/kc7_scoreboard
# Load 'Manage Users' page more efficiently

Fixes Issue #22 

## Description
The purpose of these changes is to improve the performance of the 'Manage Users' page by loading data server-side instead of client-side.

## Changes made

### app/client/templates/admin/manage_users.html

- Data is no longer gathered via a jQuery loop. Instead, the table is rendered by passing it through the route.
 ```
<div class="row">
    <div class="col-md-12">
      <div class="card shadow mb-4">
        <div class="card-header py-3">
          <h6 class="m-0 font-weight-bold text-primary">Users</h6>
        </div>
        {% block content %}
          {{ table }}
        {% endblock %}
      </div>
    </div>
  </div>
```

- Updated script to enable searching and displaying 50 users per page.
```
<script>
    $(document).ready(function() {
        $('#usersTable').DataTable({
            "paging": true, 
            "pageLength": 50, 
            "searching": true
        });
    });
  </script>
```

### app/server/models.py

- Created **UserTable** class that defines and formats the table passed to the 'Manage Users' page.
```
class UserTable(Table):
    # Define columns for the UserTable
    id = Col('ID')
    username = Col('Username')
    email = Col('Email')
    is_admin = Col('Admin?')
    delete = Col('Delete')

    def __init__(self, items):
        super().__init__(items)

        # Set td_contents for delete column
        self.delete.td_contents = lambda item, attr_list: Markup(
            f'<form action="/deluser" method="post">'
            f'<input type="hidden" name="user_id" value="{item.id}">'
            f'<input type="submit" value="Delete" '
            f'onclick="return confirm(\'Are you sure you want to delete this user?\')" '
            f'class="btn btn-danger btn-block btn-sm"></form>'
        )

        # Set td_contents for is_admin column
        self.is_admin.td_contents = lambda item, attr_list: Markup(
            f'<form action="/toggle_admin" method="post">'
            f'<div class="form-check form-switch">'
            f'<input onChange="this.form.submit()" data-toggle="toggle" data-onstyle="primary" '
            f'class="form-check-input" type="checkbox" name="is_admin" '
            f'data-on="Admin" data-off="Not Admin" id="flexSwitchCheckDefault" '
            f'{"checked" if item.is_admin else ""}>'
            f'<input type="hidden" name="user_id" value="{item.id}"></div></form>'
        )

    def __html__(self):
        inner_html = super().__html__()

        # Wrap the inner HTML with a table element and return it
        return f'<table class="table" id="usersTable">{inner_html[7:-8]}</table>'
```

### app/server/views.py

- Updated query to only send necessary data to the 'Manage Users' page. The data is then filtered via the UserTable class and rendered with HTML.

```
user_list = db.session.query(
        Users.id, Users.username, Users.email,
        case([(Roles.name == 'Admin', True)], else_=False).label('is_admin')
    ).select_from(Users).outerjoin(UserRoles).outerjoin(Roles).group_by(
        Users.id, Users.username, Users.email, Roles.name
    ).all()

    table = UserTable(user_list)

    return render_template("admin/manage_users.html", table=table)`
```

## How to test

1. To test the new loading speed with a large dataset, switch to the "prod" database.
2. Navigate to the leftmost navbar and select "Admin Central."
3. From the dropdown menu, select "Manage Users."
4. Initially, you may see unordered data for a brief moment until the page finishes rendering with the table specifications specified in the "manage_users.html" script.
